api/tools/src/main/java/org/pmiops/workbench/tools/PopulateOpsUserAffiliations.java

```diff
@@ -0,0 +1,209 @@
+                dryRun,
+                String.format(
+                    "Saved AouOps Institutional Affiliation for %s",
+                    affiliation.getUser().getUsername()));
+          });
+    };
+  }
+
+  private static void dryLog(boolean dryRun, String msg) {
+    String prefix = "";
+    if (dryRun) {
+      prefix = "[DRY RUN] Would have... ";
+    }
+    log.info(prefix + msg);
+  }
+
+  public static void main(String[] args) {
+    CommandLineToolConfig.runCommandLine(PopulateOpsUserAffiliations.class, args);
+  }
+}
+
+/**
+ * Structure the incoming ops user data according to the format of the source CSV:
+ *
+ * <p>First Name,Last Name,Email,"Workbench Email",Institution,Role,Action
+ */
+class OpsUser {
+  final String firstName;
+  final String lastName;
+  final String contactEmail; // "Email" to DRC admin staff
+  final String userName; // "Workbench Email" to DRC admin staff
+  final String operationalRole;
+  final String action;
+
+  static final int COLUMNS = 6;
+
+  private static final Logger log = Logger.getLogger(OpsUser.class.getName());
+
+  private OpsUser(final String[] userLine) {
+    this.firstName = userLine[0].trim();
+    this.lastName = userLine[1].trim();
+    this.contactEmail = userLine[2].trim();
+    this.userName = userLine[3].trim();
+    this.operationalRole = userLine[4].trim();
+    this.action = userLine[5].trim();
+  }
+
+  static List<OpsUser> parseInput(final String filename) throws IOException {
+    try (final CSVReader reader = new CSVReader(new FileReader(filename))) {
+      // consume and sanity-check header line
+      final String[] headerLine = reader.readNext();
+      if (headerLine.length != COLUMNS) {
+        throw new RuntimeException(
+            String.format(
+                "Expected %d columns in input file. Was: %d", COLUMNS, headerLine.length));
+      }
+
+      return StreamSupport.stream(reader.spliterator(), false)
+          .map(OpsUser::new)
+          .collect(Collectors.toList());
+    }
+  }
+
+  private void checkField(String dbValue, String csvValue, String fieldName) {
+    if (!dbValue.equals(csvValue)) {
+      log.warning(
+          String.format(
+              "CSV and DB values do not match for user '%s', field '%s'. CSV = %s, DB = %s",
+              userName, fieldName, csvValue, dbValue));
+    }
+  }
+
+  public DbUser dbCheck(UserDao userDao, VerifiedInstitutionalAffiliationDao affiliationDao) {
+    final DbUser dbUser = userDao.findUserByUsername(userName);
+
+    if (dbUser == null) {
+      throw new RuntimeException(String.format("User %s was not found in the DB", userName));
+    }
+
+    checkField(dbUser.getGivenName(), firstName, "First Name");
+    checkField(dbUser.getFamilyName(), lastName, "Last Name");
+    checkField(dbUser.getContactEmail(), contactEmail, "Contact Email");
+
+    affiliationDao
+        .findFirstByUser(dbUser)
+        .ifPresent(
+            affiliation -> {
+              throw new RuntimeException(
```

Comment:
I recommend you don't throw for one invalid input item in a batch job. Log an error and/or write a structured results file to another CSV file (or for complex migrations, to a temporary migration job staging table).

Reply:
It would be very strange for this to be true.  I want us to fail hard in that case.  I can add a comment.